### ISSUE TITLE
Using a standalone Validator class

### ISSUE THREAD

**Issue by LukeDornon** [formio/formio#1828]
**Describe the bug**
Calls to Validator.validate are apparently creating global objects that gc never removes.

**Version/Branch**
3.5.5

**To Reproduce**
I have an express.js application that validates managed form schemas before storage in database tables.  It exposes an endpoint to post the schema and value then creates a validator to validate.  Depending on the size of the schema and value payload, the memory allocation grows very rapidly.

**Screenshots**
Validator call:
![image](https://github.com/user-attachments/assets/276dcffd-406a-40e0-bbbd-2fdc87ce4c8e)

GC trace logs show memory allocation rising:
![image](https://github.com/user-attachments/assets/ee5d858a-c1f0-4225-bc0e-a479e2c19db3)

**Additional context**
Running in node.js v16.18.0.

**Comment by lane-formio** [comment]
Have you tested any of the 4.x versions to see if this is still reproducible?

**Comment by LukeDornon** [comment]
I tried v4.0.0 and v4.1.0 but I hit "TypeError: Cannot read properties of undefined (reading 'x-jwt-token')" when I call the validator constructor.  It's not immediately clear to in the validator.js source where this is happening.

**Comment by LukeDornon** [comment]
I updated the validator constructor as shown, but I can't find any description of what's missing. 
![image](https://github.com/user-attachments/assets/78ccf7f5-1172-4fc3-b698-1f02db84c9dd)

**Comment by lane-formio** [comment]
<div>Agent Savvas Sidiropoulos linked Freshdesk ticket <a href="https://formhelp.freshdesk.com/a/tickets/51357">51357</a> for this issue.</div><br/>

**Comment by brendanbond** [comment]
@LukeDornon yeah, we can look into this. It would help us get to the bottom of the issue if you could provide a simple toy example of your API, something like a CodeSandbox so we can understand (a) why the constructor is giving you a runtime error and (b) whether/why there's a memory leak.

**Comment by LukeDornon** [comment]
I created two forks of a sandbox, one using v3.5.5 and one using v4.1.0 to demonstrate the two problems. https://codesandbox.io/p/devbox/unruffled-architecture-8gsvpn
https://codesandbox.io/p/devbox/wild-butterfly-7zq7yg

My application would post a payload something like this:
{
    "schema": {
        "display": "form",
        "components": [{
                "label": "thing 1",
                "tableView": true,
                "key": "thing1",
                "type": "textfield",
                "input": true
            }, {
                "label": "thing 2",
                "tableView": true,
                "key": "thing2",
                "type": "textfield",
                "input": true
            }
        ]
    },
    "value": {
        "thing1": "thing",
        "thing2": "thing"
    }
}

In the first case it returns success but leaks memory, the second case immediately fails for failure to read properties of undefined.

**Comment by brendanbond** [comment]
@LukeDornon this is so helpful, really appreciate it

**Comment by lane-formio** [comment]
I wanted to note that we expect to resource a dev to look into both of these issues.

For internal reference: FIO-9364 & FIO-9365

**Comment by brendanbond** [comment]
hey @LukeDornon #1833 should fix your runtime error with the Validator class. We've done some testing and don't see a memory leak issue with 4.x - let us know what you find.

**Comment by LukeDornon** [comment]
Hi @brendanbond, I installed the 1833 rc but I am seeing this error when I attempt to run the client.

![image](https://github.com/user-attachments/assets/e500f069-9d09-4e76-ae88-233a78f6ec68)


Testing out older versions it appears I can run up to 4.2.0 but starting in 4.2.1 the error appears.

**Comment by LukeDornon** [comment]
I did attempt a workaround dynamically importing the validator class (const Validator = dynamic(() => import("formio/src/resources/Validator.js"), { ssr: false });) but that then throws the typeError Validator is not a constructor when instantiating the class.

**Comment by brendanbond** [comment]
We haven't published an rc of this change, you'll need to use the CI generated npm version: 4.0.0-dev.1833.5619ce6

Does this fix the issue?

**Comment by LukeDornon** [comment]
No, it happens in any version I've tried newer than 4.2.0.

**Comment by brendanbond** [comment]
1. I got this to work by (a) linking the latest releases of @formio/core and @formio/vm to my toy application and (b) making some minor changes to runtime execution in the Validator class. I'll try and update the package.json file for this library early next week to reflect these changes, and an official release will be coming pretty soon.

2. As validation has gotten quite a bit more complex, it strikes me that using the Validator as a standalone singleton might not be the right direction to go. Is there any reason you wouldn't just use the server as it is, and use the `/validate` endpoint? To be perfectly honest, I'm not this class was designed to be exported and used in a standalone application, it depends on a lot of formio server-specific architecture to do its thing.

**Comment by LukeDornon** [comment]
Any update on a release to support the validator in v4?

**Comment by lane-formio** [comment]
@LukeDornon 
We believe to be resolved in 4.3.0, we were not able to reproduce any memory leaks there.

**Comment by LukeDornon** [comment]
Are you able to detail the steps you took to get 4.3.0 working?  My
application is still throwing 'TypeError: Cannot read properties of
undefined (reading 'x-jwt-token')' when instantiating the validator.  I did
include @formio/core and @formio/vm in the project.

**Comment by brendanbond** [comment]
Good news and bad news here.

The good news is that getting it to work is pretty simple, you just need to make your own spoofed express request object. Here's my toy app handler versus formio v4.3.0:
```javascript
app.post("/", (req, res) => {
  try {
    const hook = {
      alter: () => {},
    };
    const myReq = {
      ...req,
      headers: {},
      currentForm: req.body.schema,
    };
    const validator = new Validator(
      myReq,
      undefined,
      undefined,
      undefined,
      hook
    );
    validator.validate({data: req.body.value}, (results) => {
      if (results) {
        res.status(400).json(results.details);
      } else {
        res.status(200).send();
      }
    });
  } catch (ex) {
    console.error(ex);
    res.status(500).json(ex);
  }
  req = undefined;
});
```

The bad news: we've _already_ got changes coming down the pipe to the Validator function signature (there are a couple of arguments being added in a version that will be released any day now), and there is exactly zero guarantee around that function signature's stability. Additionally, we've been adding `exports` fields to our libraries' package.json files to prevent this type of explicit deep import (i.e. you won't be able to require directly from `src` e.g. `require('formio/src/resources/Validator.js')`, and although there are no plans to do so with this library for the moment I can't guarantee that will be always be the case. You may wake up one day and not be able to import the Validator class. My advice is to check out [@formio/core](https://github.com/formio/core), which is what the Validator class uses [under the hood](https://github.com/formio/formio/blob/ce2490179e33f8698c54c6ba46ed6fdf066d939f/src/resources/Validator.js#L297) - it's our new isomorphic data processing system (so validation, calculated values, conditionally hidden fields, all the rest). It's certainly not perfect and it is a work in progress, but it's in use today in our enterprise product and it's a far more stable option than a deep import from the `src` directory of this library. I'm closing this issue for now, but feel free to let us know if you have any more questions or concerns.

**Comment by LukeDornon** [comment]
Thank you, 4.3.0 also fixes the memory leak in my testing.
